diff -ur a/gdb/remote.c b/gdb/remote.c
--- a/gdb/remote.c	2014-07-29 20:37:42.000000000 +0800
+++ b/gdb/remote.c	2021-01-04 14:07:42.186554234 +0800
@@ -6062,8 +6062,21 @@
   buf_len = strlen (rs->buf);

   /* Further sanity checks, with knowledge of the architecture.  */
-  if (buf_len > 2 * rsa->sizeof_g_packet)
-    error (_("Remote 'g' packet reply is too long: %s"), rs->buf);
+  //if (buf_len > 2 * rsa->sizeof_g_packet)
+  //  error (_("Remote 'g' packet reply is too long: %s"), rs->buf);
+
+  if (buf_len > 2 * rsa->sizeof_g_packet) {
+      rsa->sizeof_g_packet = buf_len ;
+      for (i = 0; i < gdbarch_num_regs (gdbarch); i++) {
+          if (rsa->regs->pnum == -1)
+             continue;
+          if (rsa->regs->offset >= rsa->sizeof_g_packet)
+             rsa->regs->in_g_packet = 0;
+          else
+             rsa->regs->in_g_packet = 1;
+      }
+  }
+

   /* Save the size of the packet sent to us by the target.  It is used
      as a heuristic when determining the max size of packets that the
